transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,598 @@
+                let connection = Connection::from_quinn_connection(
+                    self.channel.clone(),
+                    connection,
+                    connection_id,
+                    rx,
+                );
+                self.alive_connections.insert(connection_id, tx);
+                let _ = result.send(Ok(connection));
+            }
+
+            // A connection wants to notify the endpoint of something.
+            ToEndpoint::ProcessConnectionEvent {
+                connection_id,
+                event,
+            } => {
+                let has_key = self.alive_connections.contains_key(&connection_id);
+                if !has_key {
+                    return ControlFlow::Continue(());
+                }
+                // We "drained" event indicates that the connection no longer exists and
+                // its ID can be reclaimed.
+                let is_drained_event = event.is_drained();
+                if is_drained_event {
+                    self.alive_connections.remove(&connection_id);
+                    if self.is_orphaned && self.alive_connections.is_empty() {
+                        tracing::info!(
+                            "Listener closed and no active connections remain. Shutting down the background task."
+                        );
+                        return ControlFlow::Break(());
+                    }
+                }
+
+                let event_back = self.endpoint.handle_event(connection_id, event);
+
+                if let Some(event_back) = event_back {
+                    debug_assert!(!is_drained_event);
+                    if let Some(sender) = self.alive_connections.get_mut(&connection_id) {
+                        // We clone the sender to guarantee that there will be at least one
+                        // free slot to send the event.
+                        // The channel can not grow out of bound because an `event_back` is
+                        // only sent if we prior received an event from the same connection.
+                        // If the connection is busy, it won't sent us any more events to handle.
+                        let _ = sender.clone().start_send(event_back);
+                    } else {
+                        tracing::error!("State mismatch: event for closed connection");
+                    }
+                }
+            }
+
+            // Data needs to be sent on the UDP socket.
+            ToEndpoint::SendUdpPacket(transmit) => self.send_packet_out(transmit),
+        }
+        ControlFlow::Continue(())
+    }
+
+    /// Handle datagram received on the socket.
+    /// The datagram content was written into the `socket_recv_buffer`.
+    fn handle_datagram(&mut self, bytes: Vec<u8>, packet_src: SocketAddr) -> ControlFlow<()> {
+        let packet = From::from(bytes.as_bytes());
+        let local_ip = self.channel.socket_addr.ip();
+        // TODO: ECN bits aren't handled
```

Comment:
They are handled in `quinn-udp` crate.

Reply:
Yes. But integrating `quinn-udp` quite complex and we'd more or less have to copy the code for that from `quinn`. Given that this feature isn't strictly needed in the first version I would stick with the `std` sockets for now. See https://github.com/libp2p/rust-libp2p/issues/2883#issuecomment-1259397727. We can change this later in follow-up PRs. Also, if we decide long-term to use `quinn` directly,  that would already be solved